unified_planning/model/ma_problem.py

```diff
@@ -0,0 +1,429 @@
+        s.append("]\n\n")
+
+        s.append("agents actions = [\n")
+        for ag in self.agents:
+            for a in ag.actions:
+                s.append(f"  {str(ag._name)}: {str(a)}\n")
+        s.append("]\n\n")
+
+        if len(self.user_types) > 0:
+            s.append("objects = [\n")
+            for ty in self.user_types:
+                s.append(f"  {str(ty)}: {str(list(self.objects(ty)))}\n")
+            s.append("]\n\n")
+
+        s.append("initial fluents default = [\n")
+        for ag in self.agents:
+            for f in ag._fluents:
+                if f in ag._fluents_defaults:
+                    v = ag._fluents_defaults[f]
+                    s.append(f"  {str(ag._name)}: {str(f)} := {str(v)}\n")
+        s.append("]\n\n")
+
+        s.append("agents initial values = [\n")
+        for ag, initial_values in self.agents_initial_values.items():
+            for initial_value in initial_values:
+                for k, v in initial_value.items():
+                    s.append(f" {str(ag._name)}: {str(k)} := {str(v)}\n")
+        s.append("]\n\n")
+
+        s.append("MA-environment initial values = [\n")
+        for k, v in self._env_initial_value.items():
+            s.append(f" MA-environment: {str(k)} := {str(v)}\n")
+        s.append("]\n\n")
+
+        s.append("agents goals = [\n")
+        for ag, goal in self.agents_goals.items():
+            s.append(f" {str(ag._name)}: {str(goal)}\n")
+        s.append("]\n\n")
+
+        s.append("MA-environment goals = [\n")
+        if len(self._env_goals) > 0:
+            for g in self._env_goals:
+                s.append(f"  {'MA-Environment'} {str(g)}\n")
+        s.append("]\n\n")
+
+        return "".join(s)
+
+
+    def __hash__(self) -> int:
+        res = hash(self._kind) + hash(self._name)
+        for ag in self.agents:
+            for f in ag._fluents:
+                res += hash(f)
+        for ag in self.agents:
+            for a in ag._actions:
+                res += hash(a)
+        for ut in self._user_types:
+            res += hash(ut)
+        for o in self._objects:
+            res += hash(o)
+        for iv in self._env_initial_value.items():
+            res += hash(iv)
+        for goals in self._agent_goals.values():
+            for g in goals:
+                res += hash(g)
+        return res
+
+    def has_name(self, name: str) -> bool:
+        """Returns true if the name is in the problem."""
+        return (
+            self.get_ma_environment.has_fluent(name)
+            or self.has_object(name)
+            or self.has_type(name)
+            or self.has_agent(name)
+        )
+
+    def add_environment(self, env_ma) -> Union["up.model.environment_ma.MAEnvironment"]:
+        """Add a MAenvironment."""
+        self._env_ma = env_ma
+        return self._env_ma
+
+    @property
+    def get_ma_environment(self) -> Union["up.model.environment_ma.MAEnvironment"]:
```

Comment:
`Union` is used when you want to represent more than one type
```suggestion
    def get_ma_environment(self) -> "up.model.environment_ma.MAEnvironment":
```

Reply:
Ok!